cuequivariance_torch/cuequivariance_torch/primitives/segmented_polynomial_product.py

```diff
@@ -0,0 +1,274 @@
+import torch
+import torch.nn as nn
+from typing import List, Optional, Dict
+from itertools import accumulate
+
+try:
+    # keep us an option to be independent of the torch.library machinery
+    from cuequivariance_ops_torch.tensor_product_uniform_1d_jit import (
+        tensor_product_uniform_1d_jit,
+        BATCH_DIM_SHARED,
+        BATCH_DIM_BATCHED,
+        BATCH_DIM_INDEXED,
+        BATCH_DIM_AUTO,
+    )
+except Exception:
+    from cuequivariance_ops_torch.tensor_product_uniform_1d_jit import (
+        BATCH_DIM_SHARED,
+        BATCH_DIM_BATCHED,
+        BATCH_DIM_INDEXED,
+        BATCH_DIM_AUTO,
+    )
+
+    def tensor_product_uniform_1d_jit(
+        name: str,
+        math_dtype: torch.dtype,
+        operand_extent: int,
+        num_inputs: int,
+        num_outputs: int,
+        num_index: int,
+        buffer_dim: List[int],
+        buffer_num_segments: List[int],
+        batch_dim: List[int],
+        index_buffer: List[int],
+        dtypes: List[int],
+        num_operations: int,
+        num_operands: List[int],
+        operations: List[int],
+        num_paths: List[int],
+        path_indices_start: List[int],
+        path_coefficients_start: List[int],
+        path_indices: List[int],
+        path_coefficients: List[float],
+        batch_size: int,
+        tensors: List[torch.Tensor],
+    ) -> List[torch.Tensor]:
+        return torch.ops.cuequivariance_ops.tensor_product_uniform_1d_jit(
+            name,
+            math_dtype,
+            operand_extent,
+            num_inputs,
+            num_outputs,
+            num_index,
+            buffer_dim,
+            buffer_num_segments,
+            batch_dim,
+            index_buffer,
+            dtypes,
+            num_operations,
+            num_operands,
+            operations,
+            num_paths,
+            path_indices_start,
+            path_coefficients_start,
+            path_indices,
+            path_coefficients,
+            batch_size,
+            tensors,
+        )
+
+
+class SegmentedPolynomialProductJit(nn.Module):
+    def __init__(
+        self,
+        polynomial,
+        math_dtype: torch.dtype = torch.float32,
+        output_dtype_map: List[int] = None,
+        name: str = "segmented_polynomial",
+    ):
+        super().__init__()
+
+        operand_extent = None
+        for o in polynomial.operands:
+            torch._assert(
+                o.ndim in [0, 1], "only 0 or 1 dimensional operands are supported"
+            )
+            torch._assert(
+                all(len(s) == o.ndim for s in o.segments),
+                "all segments must have the same number of dimensions as the operand",
+            )
```

Comment:
side comment: this is supposed to be always true unless there is a bug elsewhere. But it's good to keep as a sanity check